feat(queryAPI): add filter for instances with jobs retrying in java API

PHWaechtler opened the pull request:
related to #4904

psavidis commented:
### Review

- 🔧 For the sake of uniformity and [Testing Best Practices](https://github.com/camunda/camunda-bpm-platform/wiki/Testing-Best-Practices-Java#:~:text=Use%20fluent%20AssertJ%20assertions) (favouring the DSL explicitness to avoid ambiguity), AssertJ also for the equality check. I've did collected the suggestions here to avoid repeating the same point over and over again:
- Unfortunately most classes in the project are not formatted entirely. The changes in the [HistoricProcessInstanceTest](https://github.com/camunda/camunda-bpm-platform/blob/0728abd1daae9fd26ddd233cd2973dc61b12d68d/engine/src/test/java/org/camunda/bpm/engine/test/history/HistoricProcessInstanceTest) span across the whole class due to the formatter. This makes the discovery of the changes to the file hard to track. 
❓ Can we get rid of the formatting changes to the file so that only the changes for this feature ?
    - I assume the real change is only the addition of `testHistoricProcessInstanceQueryWithJobsRetrying` ?

PHWaechtler commented:
> **1. AssertJ**
> 
> * 🔧 For the sake of uniformity and [Testing Best Practices](https://github.com/camunda/camunda-bpm-platform/wiki/Testing-Best-Practices-Java#:~:text=Use%20fluent%20AssertJ%20assertions) (favouring the DSL explicitness to avoid ambiguity), AssertJ also for the equality check. 

I would also prefer AssertJ, but the rest of the test cases uses JUni. For consistency I then also used the same style assertions as in the other test cases, should I rewrite the rest of the test cases in this class too?

> 
> **2. Formatting** Unfortunately most classes in the project are not formatted entirely. The changes in the [HistoricProcessInstanceTest](https://github.com/camunda/camunda-bpm-platform/blob/0728abd1daae9fd26ddd233cd2973dc61b12d68d/engine/src/test/java/org/camunda/bpm/engine/test/history/HistoricProcessInstanceTest) span across the whole class due to the formatter. This makes the discovery of the changes to the file hard to track.

Yeah I brought this up a couple of times as I was susprised to see the lack of formatting, this will always bloat PRs in annoying ways. I am considering formatting files that a PR will touch first in a separate commit before adding changes, but even that is kinda inconvenient when working as you cant always predict all files you're about to change. I've not found another workaround if we dont want to fix the formatting in the codebase. 

> ❓ Can we get rid of the formatting changes to the file so that only the changes for this feature show up in the PR? I assume the real change is only the addition of `testHistoricProcessInstanceQueryWithJobsRetrying` ?

If we undo the formatting changes the file remains unformatted though, introducing untidy code and not fixing the current formatting issues. I feel like that will just making be the issue worse 🤔 


> * ❓  **New SQL Condition Coverage**: The newly added subquery has two conditions on `EXCEPTION_MSG` and the `EXCEPTION_STACK_ID`. Since the test uses the [FailingDelegate](https://github.com/camunda/camunda-bpm-platform/blob/0728abd1daae9fd26ddd233cd2973dc61b12d68d/engine/src/test/resources/org/camunda/bpm/engine/test/api/history/testInstancesWithJobsRetrying.bpmn20.xml#L12) to simulate the job failure, would it make sense to control the [FailingDelegate#message](https://github.com/camunda/camunda-bpm-platform/blob/8f7940067e238c9a49c1239a32925931771c57ef/engine/src/test/java/org/camunda/bpm/engine/test/api/runtime/FailingDelegate.java#L31) variable and add a test for :
>   
>   * A Failing Delegate with Null message
>   * A Failing Delegate with a `ProcessEngineException` with the empty constructor
>

Yes, I oriented the conditions on the existing Job Query for this type of scenario, covering the same fields. I do not know how to trigger the two scnearios you've listed, do you have a pointer to where similar behaviour has been used in a test case so I can refer to that for implemenation?

> **4. CI** Since the changes add the new criterion to the Query API and the change involves changing the SQL, building the effective PR changes against the following two profiles will increase our confidence :
> 
> * [ ]  `ci:default-build`
> * [ ]  `ci:all-db`

Sounds good will add once I added the changes.

PHWaechtler commented on engine/src/test/java/org/camunda/bpm/engine/test/history/HistoricProcessInstanceTest.java:
```diff
@@ -502,13 +545,52 @@ public void testHistoricProcessInstanceQueryWithIncidentMessageLikeNull() {
+    runtimeService.startProcessInstanceByKey("processWithJobsRetrying");
+
+    // then
+    assertEquals(0, queryWithJobsRetrying.count());
```
Now that I know I am allowed to use assertJ then I would prefer `.isZero()` for all of these :)

psavidis replied:
Let's do that :)

PHWaechtler commented on engine/src/test/java/org/camunda/bpm/engine/test/history/HistoricProcessInstanceTest.java:
```diff
@@ -502,13 +545,52 @@ public void testHistoricProcessInstanceQueryWithIncidentMessageLikeNull() {
+    // then
+    assertEquals(1, queryWithJobsRetrying.count());
+    assertEquals(1, queryWithJobsRetrying.list().size());
+    assertEquals(instanceWithRetryingJob.getId(), queryWithJobsRetrying.singleResult().getRootProcessInstanceId());
```
> when you query for retrying jobs you should find the instance with retrying jobs.
I am expecting to find instances because we are querying for instances that are associated with retrying jobs. Is this Job ID a field of the instance you're referring to? I am not seeing it. Or did you mean instance ID? I assumed that was the same as rootInstanceID just by looking at the values/name, didnt realise there was additional logic behind that. I have changed it to just the ID of the instance. Lmk if it should be pointing to some job though

PHWaechtler commented:
A summary of the changes from code review:

**AssertJ**
Adjusted the assertions in the new test case but left the others as is. Do we have a ticket for this kind of maintenance, or should we open a follow up? It looks like there would be a lot of tests to adjust.

**Formatting**
Discussed our dislike of the current state and I will be more wary to try to move the auto formatting to separate commits in following PRs. Plus might push for this topic on a team level again, but its outside of the scope of this ticket.

**SQL condition coverage for null exception msg/stack trace**
We discussed this on zoom and tried it, but it seems the effort to try to create this scenario is not worth it. We also discussed that this instance query should in behaviour match the existing job query for jobs with exceptions, the existing job query has the same sql, giving us more confidence.

**CI**
Added the additional labels `ci:default-build` and `ci:all-db`
